spock-core/src/main/groovy/spock/util/SourceToAstNodeAndSourceTranspiler.groovy

```diff
@@ -0,0 +1,1241 @@
+    print ' : '
+    expression?.falseExpression?.visit this
+  }
+
+  @Override
+  void visitShortTernaryExpression(ElvisOperatorExpression expression) {
+    visitTernaryExpression(expression)
+  }
+
+  @Override
+  void visitBooleanExpression(BooleanExpression expression) {
+    expression?.expression?.visit this
+  }
+
+  @Override
+  void visitWhileLoop(WhileStatement statement) {
+    printStatementLabels(statement)
+    print 'while ('
+    statement?.booleanExpression?.visit this
+    print ') {'
+    printLineBreak()
+    indented {
+      statement?.loopBlock?.visit this
+    }
+    printLineBreak()
+    print '}'
+    printLineBreak()
+  }
+
+  @Override
+  void visitDoWhileLoop(DoWhileStatement statement) {
+    printStatementLabels(statement)
+    print 'do {'
+    printLineBreak()
+    indented {
+      statement?.loopBlock?.visit this
+    }
+    print '} while ('
+    statement?.booleanExpression?.visit this
+    print ')'
+    printLineBreak()
+  }
+
+  @Override
+  void visitCatchStatement(CatchStatement statement) {
+    print 'catch ('
+    visitParameters([statement.variable])
+    print ') {'
+    printLineBreak()
+    indented {
+      statement.code?.visit this
+    }
+    print '}'
+    printLineBreak()
+  }
+
+  @Override
+  void visitBitwiseNegationExpression(BitwiseNegationExpression expression) {
+    print '~('
+    expression?.expression?.visit this
+    print ') '
+  }
+
+  @Override
+  void visitAssertStatement(AssertStatement statement) {
+    print 'assert '
+    statement?.booleanExpression?.visit this
+    print ' : '
+    statement?.messageExpression?.visit this
+  }
+
+  @Override
+  void visitClosureListExpression(ClosureListExpression expression) {
+    boolean first = true
+    expression?.expressions?.each {
+      if (!first) {
+        print ';'
+      }
+      first = false
+      it.visit this
+    }
+  }
+
+  @Override
+  void visitMethodPointerExpression(MethodPointerExpression expression) {
+    expression?.expression?.visit this
+    trimSpaceRight() // remove space inserted by previous expression
+    print '.&'
+    expression?.methodName?.visit this
+  }
+
+  @Override
+  void visitMethodReferenceExpression(MethodReferenceExpression expression) {
+    expression?.expression?.visit this
+    print '::'
+    expression?.methodName?.visit this
+  }
+
+  @Override
+  void visitArrayExpression(ArrayExpression expression) {
+    print 'new '
+    visitType expression?.elementType
+    print '['
+    visitExpressionsAndCommaSeparate(expression?.sizeExpression)
+    print ']'
+    if (expression?.expressions) { // print array initializer
+      print '{'
+      visitExpressionsAndCommaSeparate(expression?.expressions)
+      print '}'
+    }
+  }
+
+  private void visitExpressionsAndCommaSeparate(List<? super Expression> expressions) {
+    boolean first = true
+    expressions?.each {
+      if (!first) {
+        print ', '
+      }
+      first = false
+      ((ASTNode)it).visit this
```

Comment:
I have no idea why the `AnnotationConstantExpression` does it, but it first visits the values of the members before it visits itself. That's why you got `org.spockframework.runtime.model.BlockKind.SETUP` followed by `[]` followed by the actual correct representation.
```suggestion
      if (it instanceof AnnotationConstantExpression) {
        visitConstantExpression(it)
      } else {
        (it as ASTNode).visit this
      }
```

Reply:
Reported it as https://issues.apache.org/jira/browse/GROOVY-9980